jerry-debugger/src/lib/protocol-handler.ts

```diff
@@ -0,0 +1,251 @@
+  [cp: string]: ParsedFunction;
+}
+
+// abstracts away the details of the protocol
+export class JerryDebugProtocolHandler {
+  public delegate: JerryDebugProtocolDelegate;
+
+  // debugger configuration
+  private maxMessageSize: number = 0;
+  private byteConfig: ByteConfig;
+  private version: number = 0;
+  private functionMap: ProtocolFunctionMap;
+
+  private stack: Array<ParserStackFrame>;
+  private source: string = '';
+  private sourceData?: Uint8Array;
+  private sourceName?: string;
+  private sourceNameData?: Uint8Array;
+  private functionName?: Uint8Array;
+  private functions: FunctionMap = {};
+  private lastScriptID: number = 0;
+
+  constructor(delegate: JerryDebugProtocolDelegate) {
+    this.delegate = delegate;
+
+    this.byteConfig = {
+      cpointerSize: 0,
+      littleEndian: true,
+    };
+
+    this.functionMap = {
+      [SP.JERRY_DEBUGGER_CONFIGURATION]: this.onConfiguration,
+      [SP.JERRY_DEBUGGER_BYTE_CODE_CP]: this.onByteCodeCP,
+      [SP.JERRY_DEBUGGER_PARSE_FUNCTION]: this.onParseFunction,
+      [SP.JERRY_DEBUGGER_BREAKPOINT_LIST]: this.onBreakpointList,
+      [SP.JERRY_DEBUGGER_BREAKPOINT_OFFSET_LIST]: this.onBreakpointList,
+      [SP.JERRY_DEBUGGER_SOURCE_CODE]: this.onSourceCode,
+      [SP.JERRY_DEBUGGER_SOURCE_CODE_END]: this.onSourceCode,
+      [SP.JERRY_DEBUGGER_SOURCE_CODE_NAME]: this.onSourceCodeName,
+      [SP.JERRY_DEBUGGER_SOURCE_CODE_NAME_END]: this.onSourceCodeName,
+    };
+
+    this.stack = [{
+      isFunc: false,
+      line: 1,
+      column: 1,
+      name: '',
+      source: '',
+      lines: [],
+      offsets: [],
+    }];
+  }
+
+  // for the temporary mollification of lint
+  unused() {
+    this.maxMessageSize;
+  }
+
+  decodeMessage(format: string, message: Uint8Array, offset: number) {
+    return decodeMessage(this.byteConfig, format, message, offset);
+  }
+
+  onConfiguration(data: Uint8Array) {
+    console.log('[Configuration]');
+    if (data.length !== 5) {
+      this.abort('configuration message wrong size');
+      return;
+    }
+
+    this.maxMessageSize = data[1];
+    this.byteConfig.cpointerSize = data[2];
+    this.byteConfig.littleEndian = Boolean(data[3]);
+    this.version = data[4];
+
+    if (this.byteConfig.cpointerSize !== 2 && this.byteConfig.cpointerSize !== 4) {
+      this.abort('compressed pointer must be 2 or 4 bytes long');
+    }
+
+    if (this.version !== JERRY_DEBUGGER_VERSION) {
+      this.abort('incorrect target debugger version detected: ' + this.version
+                 + ' expected: ' + JERRY_DEBUGGER_VERSION);
+    }
+  }
+
+  onByteCodeCP(data: Uint8Array) {
+    console.log('[Byte Code CP]');
+    const frame = this.stack.pop();
+    if (!frame) {
+      throw new Error('missing parser stack frame');
+    }
+
+    const byteCodeCP = this.decodeMessage('C', data, 1)[0];
+    const func = new ParsedFunction(byteCodeCP, frame);
+
+    this.functions[byteCodeCP] = func;
+    if (this.stack.length > 0) {
+      return;
+    }
```

Comment:
I'm a bit fuzzy on why the `func` gets assigned `source` and `sourceName` if it was the last/bottom frame.
If other `func`s don't have these assigned, won't the `toString()` function for breakpoints in that func print `<unknown>` as sourceName?

Reply:
I'm afraid all these last few questions I'm also a bit fuzzy on as they just come from the way that the original HTML debugger was written. But they did serve some purpose. I'd prefer to progress further in the protocol and deal with these things when I get up to working on the functionality that requires them.